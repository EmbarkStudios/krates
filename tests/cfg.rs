mod util;

use krates::cfg_expr::targets;
use util::{build, cmp};

#[test]
fn ignores_non_linux() {
    let mut kb = krates::Builder::new();
    kb.include_targets(targets::ALL_BUILTINS.iter().filter_map(|ti| {
        if ti.os == Some(targets::Os::linux) {
            Some((ti.triple.clone(), Vec::new()))
        } else {
            None
        }
    }));

    let grafs = build("all-features.json", kb).unwrap();

    let targets: Vec<_> = targets::ALL_BUILTINS
        .iter()
        .filter(|ti| ti.os == Some(targets::Os::linux))
        .collect();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[test]
fn ignores_non_tier1() {
    let mut kb = krates::Builder::new();

    let targets = vec![
        targets::get_builtin_target_by_triple("i686-pc-windows-gnu").unwrap(),
        targets::get_builtin_target_by_triple("i686-pc-windows-msvc").unwrap(),
        targets::get_builtin_target_by_triple("i686-unknown-linux-gnu").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-apple-darwin").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-pc-windows-gnu").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-pc-windows-msvc").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-unknown-linux-gnu").unwrap(),
    ];

    kb.include_targets(targets.iter().map(|ti| (ti.triple.clone(), vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[test]
fn ignores_non_wasm() {
    let mut kb = krates::Builder::new();

    let targets = vec![targets::get_builtin_target_by_triple("wasm32-unknown-unknown").unwrap()];

    kb.include_targets(targets.iter().map(|ti| (ti.triple.clone(), vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[cfg(feature = "targets")]
#[test]
fn handles_non_builtin() {
    let mut kb = krates::Builder::new();

    use krates::target_lexicon::Triple;

    let xbox: Triple = "x86_64-xboxone-windows-msvc".parse().unwrap();

    kb.include_targets(std::iter::once(("x86_64-xboxone-windows-msvc", vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    !expr.eval(|pred| match pred {
                        krates::cfg_expr::Predicate::Target(tp) => tp.matches(&xbox),
                        _ => false,
                    })
                } else {
                    "x86_64-xboxone-windows-msvc" != cfg
                }
            } else {
                false
            }
        },
    );
}
