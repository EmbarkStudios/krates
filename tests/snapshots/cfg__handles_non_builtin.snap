---
source: tests/cfg.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 (path+file:///home/jake/code/krates/tests/ws/a)" ]
    1 [ label = "crate b 0.1.0 (path+file:///home/jake/code/krates/tests/ws/b)" ]
    2 [ label = "crate bumpalo 3.11.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    3 [ label = "crate c 0.1.0 (path+file:///home/jake/code/krates/tests/ws/c)" ]
    4 [ label = "crate cc 1.0.73 (git+https://github.com/alexcrichton/cc-rs#53fb72c87e5769a299f1886ead831901b9c775d6)" ]
    5 [ label = "crate cc 1.0.73 (registry+https://github.com/rust-lang/crates.io-index)" ]
    6 [ label = "crate cfg-if 1.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    7 [ label = "crate difference 2.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    8 [ label = "crate js-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    9 [ label = "crate leftpad 0.2.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    10 [ label = "crate log 0.4.17 (registry+https://github.com/rust-lang/crates.io-index)" ]
    11 [ label = "crate once_cell 1.15.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    12 [ label = "crate proc-macro2 1.0.44 (registry+https://github.com/rust-lang/crates.io-index)" ]
    13 [ label = "crate quote 1.0.21 (registry+https://github.com/rust-lang/crates.io-index)" ]
    14 [ label = "crate ring 0.16.20 (registry+https://github.com/rust-lang/crates.io-index)" ]
    15 [ label = "crate spin 0.5.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    16 [ label = "crate syn 1.0.101 (registry+https://github.com/rust-lang/crates.io-index)" ]
    17 [ label = "crate unicode-ident 1.0.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    18 [ label = "crate untrusted 0.7.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    19 [ label = "crate wasm-bindgen 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    20 [ label = "crate wasm-bindgen-backend 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    21 [ label = "crate wasm-bindgen-futures 0.4.33 (registry+https://github.com/rust-lang/crates.io-index)" ]
    22 [ label = "crate wasm-bindgen-macro 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    23 [ label = "crate wasm-bindgen-macro-support 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    24 [ label = "crate wasm-bindgen-shared 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    25 [ label = "crate winapi 0.2.8 (registry+https://github.com/rust-lang/crates.io-index)" ]
    26 [ label = "crate winapi 0.3.9 (registry+https://github.com/rust-lang/crates.io-index)" ]
    27 [ label = "feature default" ]
    28 [ label = "feature default" ]
    29 [ label = "feature default" ]
    30 [ label = "feature default" ]
    31 [ label = "feature proc-macro" ]
    32 [ label = "feature ntsecapi" ]
    33 [ label = "feature wtypesbase" ]
    34 [ label = "feature proc-macro" ]
    35 [ label = "feature spans" ]
    36 [ label = "feature default" ]
    37 [ label = "feature default" ]
    38 [ label = "feature default" ]
    39 [ label = "feature default" ]
    40 [ label = "feature default" ]
    41 [ label = "feature full" ]
    42 [ label = "feature spans" ]
    43 [ label = "feature visit" ]
    44 [ label = "feature spans" ]
    45 [ label = "feature leftpad" ]
    46 [ label = "feature leftier-strings" ]
    47 [ label = "feature lazy_static" ]
    48 [ label = "feature std" ]
    49 [ label = "feature alloc" ]
    50 [ label = "feature race" ]
    51 [ label = "feature alloc" ]
    52 [ label = "feature dev_urandom_fallback" ]
    53 [ label = "feature once_cell" ]
    54 [ label = "feature derive" ]
    55 [ label = "feature parsing" ]
    56 [ label = "feature printing" ]
    57 [ label = "feature clone-impls" ]
    58 [ label = "feature proc-macro" ]
    59 [ label = "feature quote" ]
    60 [ label = "feature spans" ]
    61 [ label = "feature std" ]
    0 -> 1 [ label = "" ]
    0 -> 27 [ label = "(dev)" ]
    1 -> 27 [ label = "" ]
    1 -> 28 [ label = "(dev) 'cfg(target_arch = \"x86_64\")'" ]
    1 -> 21 [ label = " 'cfg(all(target_vendor = \"xboxone\"))'" ]
    3 -> 4 [ label = "(build)" ]
    3 -> 29 [ label = "(dev)" ]
    3 -> 9 [ label = "" ]
    3 -> 15 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    3 -> 25 [ label = " 'cfg(target_os = \"windows\")'" ]
    8 -> 30 [ label = "" ]
    10 -> 6 [ label = "" ]
    12 -> 17 [ label = "" ]
    13 -> 12 [ label = "" ]
    13 -> 31 [ label = "" ]
    14 -> 5 [ label = "(build)" ]
    14 -> 15 [ label = " 'cfg(any(target_arch = \"x86\", target_arch = \"x86_64\", all(any(target_arch = \"aarch64\", target_arch = \"arm\"), any(target_os = \"android\", target_os = \"fuchsia\", target_os = \"linux\"))))'" ]
    14 -> 18 [ label = "" ]
    14 -> 32 [ label = " 'cfg(target_os = \"windows\")'" ]
    14 -> 33 [ label = " 'cfg(target_os = \"windows\")'" ]
    16 -> 12 [ label = "" ]
    16 -> 31 [ label = "" ]
    16 -> 13 [ label = "" ]
    16 -> 34 [ label = "" ]
    16 -> 17 [ label = "" ]
    19 -> 6 [ label = "" ]
    19 -> 22 [ label = "" ]
    19 -> 35 [ label = "" ]
    20 -> 36 [ label = "" ]
    20 -> 10 [ label = "" ]
    20 -> 37 [ label = "" ]
    20 -> 38 [ label = "" ]
    20 -> 39 [ label = "" ]
    20 -> 40 [ label = "" ]
    20 -> 41 [ label = "" ]
    20 -> 24 [ label = "" ]
    21 -> 6 [ label = "" ]
    21 -> 8 [ label = "" ]
    21 -> 30 [ label = "" ]
    22 -> 39 [ label = "" ]
    22 -> 23 [ label = "" ]
    22 -> 42 [ label = "" ]
    23 -> 38 [ label = "" ]
    23 -> 39 [ label = "" ]
    23 -> 40 [ label = "" ]
    23 -> 43 [ label = "" ]
    23 -> 41 [ label = "" ]
    23 -> 20 [ label = "" ]
    23 -> 44 [ label = "" ]
    23 -> 24 [ label = "" ]
    36 -> 2 [ label = "" ]
    45 -> 3 [ label = "" ]
    45 -> 9 [ label = "" ]
    46 -> 3 [ label = "" ]
    46 -> 45 [ label = "" ]
    47 -> 3 [ label = "" ]
    27 -> 3 [ label = "" ]
    27 -> 45 [ label = "" ]
    29 -> 7 [ label = "" ]
    37 -> 11 [ label = "" ]
    37 -> 48 [ label = "" ]
    48 -> 11 [ label = "" ]
    48 -> 49 [ label = "" ]
    49 -> 11 [ label = "" ]
    49 -> 50 [ label = "" ]
    50 -> 11 [ label = "" ]
    31 -> 12 [ label = "" ]
    38 -> 12 [ label = "" ]
    38 -> 31 [ label = "" ]
    34 -> 13 [ label = "" ]
    34 -> 31 [ label = "" ]
    39 -> 13 [ label = "" ]
    39 -> 34 [ label = "" ]
    28 -> 14 [ label = "" ]
    28 -> 51 [ label = "" ]
    28 -> 52 [ label = "" ]
    52 -> 14 [ label = "" ]
    52 -> 53 [ label = "" ]
    53 -> 14 [ label = "" ]
    53 -> 11 [ label = "" ]
    51 -> 14 [ label = "" ]
    43 -> 16 [ label = "" ]
    41 -> 16 [ label = "" ]
    40 -> 16 [ label = "" ]
    40 -> 54 [ label = "" ]
    40 -> 55 [ label = "" ]
    40 -> 56 [ label = "" ]
    40 -> 57 [ label = "" ]
    40 -> 58 [ label = "" ]
    58 -> 16 [ label = "" ]
    58 -> 31 [ label = "" ]
    58 -> 34 [ label = "" ]
    57 -> 16 [ label = "" ]
    56 -> 16 [ label = "" ]
    56 -> 59 [ label = "" ]
    59 -> 16 [ label = "" ]
    59 -> 13 [ label = "" ]
    55 -> 16 [ label = "" ]
    54 -> 16 [ label = "" ]
    30 -> 19 [ label = "" ]
    30 -> 60 [ label = "" ]
    30 -> 61 [ label = "" ]
    61 -> 19 [ label = "" ]
    60 -> 19 [ label = "" ]
    60 -> 35 [ label = "" ]
    44 -> 20 [ label = "" ]
    35 -> 22 [ label = "" ]
    35 -> 42 [ label = "" ]
    42 -> 23 [ label = "" ]
    42 -> 44 [ label = "" ]
    33 -> 26 [ label = "" ]
    32 -> 26 [ label = "" ]
}

